### PR TITLE
Low latency dash support

### PR DESCRIPTION
**Pull request by CaitlinOCallaghan**
- Add Low Latency Dash (LL-DASH) support to Shaka Packager.
- Construct segments out of many chunks (moof + mdat) of data.
- Immediately upload each chunk to its corresponding segment file as soon it is created, allowing the player to read the segment while it is being generated.
- Properly construct a LL-DASH manifest that satisfies [DASH-IF standards for Low Latency](https://conformance.dashif.org/)

Test Stream with LL-DASH mode enabled
image

WIP LL-DASH Implementation Call Tree Diagram
image

image

NOTE:
- I need to address some TODOs regarding commenting/descriptions 
- I will also write unit tests once the feature implementation is verified and approved.

### CODE REVIEW

File: packager/app/packager_main.cc [hunk 1]
```diff
@@ -325,9 +325,17 @@ base::Optional<PackagingParams> GetPackagingParams() {
   ChunkingParams& chunking_params = packaging_params.chunking_params;
   chunking_params.segment_duration_in_seconds = FLAGS_segment_duration;
   chunking_params.subsegment_duration_in_seconds = FLAGS_fragment_duration;
+  chunking_params.is_low_latency_dash = FLAGS_is_low_latency_dash;
   chunking_params.segment_sap_aligned = FLAGS_segment_sap_aligned;
   chunking_params.subsegment_sap_aligned = FLAGS_fragment_sap_aligned;
 
+  if (chunking_params.is_low_latency_dash && chunking_params.subsegment_duration_in_seconds > 0) {
+    LOG(ERROR) << "Fragment duration --fragment_duration, "
+                  "cannot be specified if LL-DASH --is_low_latency_dash, "
```
Comment (joeyparrish) [inline packager/app/packager_main.cc]: nit: I'm not convinced that these commas make sense.

Also, why can't fragment_duration be used with is_low_latency_dash?  If the explanation is too long for the error log (which I would assume to be the case), then perhaps you could put an explanation in a comment just above the error log.  Then anyone researching the error would find the full explanation in the source.

File: packager/app/mpd_flags.cc [hunk 2]
```diff
@@ -75,3 +75,6 @@ DEFINE_bool(dash_force_segment_list,
             "content is huge and the total number of (sub)segment references "
             "is greater than what the sidx atom allows (65535). Currently "
             "this flag is only supported in DASH ondemand profile.");
+DEFINE_bool(is_low_latency_dash,
+            false,
+            "TODO(Caitlin): fill this out");
```
Comment (joeyparrish) [inline packager/app/mpd_flags.cc]: Totally.  Please do.  😁

File: packager/file/file.cc [hunk 3]
```diff
@@ -100,11 +101,11 @@ File* CreateUdpFile(const char* file_name, const char* mode) {
 }
 
 File* CreateHttpsFile(const char* file_name, const char* mode) {
-  return new HttpFile(HttpMethod::kPut, std::string("https://") + file_name);
+  return new HttpFile(HttpMethod::kPost, std::string("https://") + file_name);
```
Comment (joeyparrish) [inline packager/file/file.cc]: Why does this change from PUT to POST?  When I've tested this (briefly) with hard-coded auth tokens (that expire too quickly to be useful in production), the cloud services I wrote to all worked with PUT.  Do cloud storage services usually accept both POST and PUT?  Is POST specific to an auth proxy?

File: packager/file/http_file.cc [hunk 4]
```diff
@@ -297,7 +299,7 @@ void HttpFile::SetupRequest() {
   curl_easy_setopt(curl, CURLOPT_FOLLOWLOCATION, 1L);
   curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, &CurlWriteCallback);
   curl_easy_setopt(curl, CURLOPT_WRITEDATA,
-                   method_ == HttpMethod::kPut ? nullptr : &download_cache_);
+                method_ == HttpMethod::kPost ? nullptr : &download_cache_);
```
Comment (joeyparrish) [inline packager/file/http_file.cc]: Wouldn't this be the case for both POST and PUT?  It seems like the switch above implies that both methods would still be possible.

Or perhaps we would change this to `&download_cache_` if it's `kGet`, and null otherwise?

File: packager/media/public/chunking_params.h [hunk 5]
```diff
@@ -25,6 +25,9 @@ struct ChunkingParams {
   /// Setting to subsegment_sap_aligned to true but segment_sap_aligned to false
   /// is not allowed.
   bool subsegment_sap_aligned = true;
+  /// Enable LL-DASH streaming
+  /// TODO(Caitlin): Elaborate
```
Comment (joeyparrish) [inline packager/media/public/chunking_params.h]: Yes, please!

File: packager/mpd/base/representation.cc [hunk 6]
```diff
@@ -297,6 +305,15 @@ void Representation::SetPresentationTimeOffset(
   media_info_.set_presentation_time_offset(pto);
 }
 
+void Representation::SetAvailabilityTimeOffset() {
+  // Adjust the frame duration to units of seconds to match target segment duration.
+  const float frame_duration_sec = (float)frame_duration_ / (float)media_info_.reference_time_scale();
+  const float ato = mpd_options_.mpd_params.target_segment_duration - frame_duration_sec;
```
Comment (joeyparrish) [inline packager/mpd/base/representation.cc]: I thought elsewhere I had seen that ATO would be segment time minus chunk time.  Why subtract frame time instead?  Did I misunderstand?

File: packager/media/formats/mp4/low_latency_segment_segmenter.cc [hunk 7]
```diff
@@ -0,0 +1,235 @@
...
+    // Append the segment to output file if segment template is not specified.
+    file_name_ = options().output_file_name.c_str();
+  } else {
+    file_name_ = GetSegmentName(options().segment_template,
+                               sidx()->earliest_presentation_time,
+                               num_segments_, options().bandwidth);
+  }
+
+  // Create the segment file
+  segment_file_ = File::Open(file_name_.c_str(), "a");
```
Comment (joeyparrish) [inline packager/media/formats/mp4/low_latency_segment_segmenter.cc]: How does append mode work on an HTTP "file"?

File: packager/media/formats/mp4/low_latency_segment_segmenter.cc [hunk 8]
```diff
@@ -0,0 +1,235 @@
...
+
+  // point to the already open segment file for writing
+  file.reset(segment_file_);
+  if (!file) {
+    return Status(error::FILE_FAILURE,
+                  "Cannot access the file to write " + file_name_);
+  }
+
+  // Write the chunk data to the file
+  RETURN_IF_ERROR(fragment_buffer()->WriteToFile(file.get()));
```
Comment (joeyparrish) [inline packager/media/formats/mp4/low_latency_segment_segmenter.cc]: Where/how does fragment_buffer get reset?

File: packager/media/formats/mp4/low_latency_segment_segmenter.cc [hunk 9]
```diff
@@ -0,0 +1,235 @@
...
+  if (!segment_file_) {
+    return Status(error::FILE_FAILURE, "Cannot open segment file: " +
+                                            file_name_);
+  }
+
+  std::unique_ptr<BufferWriter> buffer(new BufferWriter());
+  std::unique_ptr<File, FileCloser> file;
+
+  // Point to the already open segment file for writing
+  file.reset(segment_file_);
```
Comment (joeyparrish) [inline packager/media/formats/mp4/low_latency_segment_segmenter.cc]: I was really confused about this.  Why put a persistent handle in a unique_ptr, then release it afterward?  What value is unique_ptr bringing to this?

File: packager/media/formats/mp4/low_latency_segment_segmenter.cc [hunk 10]
```diff
@@ -0,0 +1,235 @@
...
+                                  sidx()->earliest_presentation_time,
+                                  segment_duration, segment_size);
+    is_initial_chunk_in_seg_ = false;
+  }
+
+  return Status::OK;
+}
+
+Status LowLatencySegmentSegmenter::WriteChunk() {
+  DCHECK(sidx());
```
Comment (joeyparrish) [inline packager/media/formats/mp4/low_latency_segment_segmenter.cc]: sidx isn't used in this method.

File: packager/media/formats/mp4/low_latency_segment_segmenter.cc [hunk 11]
```diff
@@ -0,0 +1,235 @@
...
+      // Set necessary values for LL-DASH mpd after the first chunk has been processed.
+      muxer_listener()->OnSampleDurationReady(sample_duration());
+      muxer_listener()->OnAvailabilityOffsetReady();
+      muxer_listener()->OnSegmentDurationReady();
+      ll_dash_mpd_values_initialized_ = true;
+    }
+    // Add the current segment in the manifest. 
+    // Following chunks will be appended to the open segment file.
+    muxer_listener()->OnNewSegment(file_name_,
+                                  sidx()->earliest_presentation_time,
```
Comment (joeyparrish) [inline packager/media/formats/mp4/low_latency_segment_segmenter.cc]: style nit: misaligned with previous line's arguments

File: packager/media/formats/mp4/low_latency_segment_segmenter.h [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+  Status WriteInitialChunk();
+  Status FinalizeSegment();
+
+  uint64_t GetSegmentDuration();
+
+  std::unique_ptr<SegmentType> styp_;
+  uint32_t num_segments_;
+  bool is_initial_chunk_in_seg_ = true;
+  bool ll_dash_mpd_values_initialized_ = false;
+  File* segment_file_;
```
Comment (joeyparrish) [inline packager/media/formats/mp4/low_latency_segment_segmenter.h]: It seems to me that this member should be a unique_ptr, to be cleaned up if the Segmenter is destroyed, rather than temporarily putting the handle into a local unique_ptr in the methods that use it, where it will only be cleaned up if there's an exception thrown or an early return while the local unique_ptr is in scope.

File: packager/media/formats/mp4/low_latency_segment_segmenter.cc (hunk 8 above)
Comment (CaitlinOCallaghan) [reply packager/media/formats/mp4/low_latency_segment_segmenter.cc]: The `fragment_buffer_` is of type `BufferWriter` and lives within `Segmenter`.  The `BufferWriter` method `WriteToFile(File* file)`  clears the buffer after writing the buffer's contents to `file`. [It can be found here in buffer_writer.cc](https://github.com/google/shaka-packager/blob/f018c9a9bf6d936f3434ea32fb6a9ee42ae12b4b/packager/media/base/buffer_writer.cc#L85).

File: packager/media/formats/mp4/low_latency_segment_segmenter.cc (hunk 10 above)
Comment (CaitlinOCallaghan) [reply packager/media/formats/mp4/low_latency_segment_segmenter.cc]: Great catch! That was accidentally left over after I abstracted out the segment duration calculation to its own method `GetSegmentDuration()`.

File: packager/media/formats/mp4/low_latency_segment_segmenter.cc (hunk 9 above)
Comment (CaitlinOCallaghan) [reply packager/media/formats/mp4/low_latency_segment_segmenter.cc]: This is an excellent question - this area was something that I was struggling with earlier. I found the flushing of the `ThreadedIoFile` and `HttpFile` and their corresponding `IoCache` buffers to be quite confusing and problematic when trying to keep the remote file open and within scope. 

This was my dirty solution to get the LL-DASH feature working, which I arrived at after a lot of fiddling and iteration. In hindsight, the implementation is kind of a wack, Frankenstein product after exploring a lot of different options.

You are totally right. The unique_ptr is not needed with the implementation above 😀 !!!

---

** NOTE: For people reading this thread who are not familiar with files within Packager, below is an outline of how `File`, `IoCacheFile`, and `HttpFile` interact. I wrote it out while reviewing the section Joey pointed out and thought that someone else might fine it helpful.

---

The call tree that I outlined is that `File::Open(...)` calls `File::Create(...)`. 

`File::Create(...)` calls `File::CreateInternalFile(...)`, which will trigger the `HttpFile` constructor via `FileFactoryFunction factory_function(...)`.  The HttpFile is pointed to by `std::unique_ptr<File, FileCloser> internal_file`.

At this point, we have our HttpFile object, `internal_file`, within `File::Create(...)`. Then, the unique_ptr `internal_file` is transferred to a new `ThreadedIoFile` via `std::move()` and stored the private variable `internal_file_`. This `ThreadedIoFile` object is returned to whoever called `File::Create(...)`. 

So, the end result of `File::Create(...)` is a `File*` to a `ThreadedIoFile`, and this `ThreadedIoFile` contains `std::unique_ptr<File, FileCloser> internal_file_`, which points to a `HttpFile` in our use case.

In `LowLatencySegmentSegmenter`, the `File* segment_file_` points to the returned `ThreadedIoFile` object, providing access to the `HttpFile`.

---

File: packager/media/formats/mp4/low_latency_segment_segmenter.cc (hunk 8 above)
Comment (joeyparrish) [reply packager/media/formats/mp4/low_latency_segment_segmenter.cc]: Perfect, thanks!  I don't know all the things.  :-)

File: packager/media/formats/mp4/low_latency_segment_segmenter.cc (hunk 9 above)
Comment (joeyparrish) [reply packager/media/formats/mp4/low_latency_segment_segmenter.cc]: Should segment_file_ be a unique_ptr<File>, then, to clean it up automatically when LowLatencySegmentSegmenter is destroyed?  You can still reset() the pointer when you finalize a segment or start a new one.

File: packager/media/formats/mp4/low_latency_segment_segmenter.h (hunk 12 above)
Comment (CaitlinOCallaghan) [reply packager/media/formats/mp4/low_latency_segment_segmenter.h]: I totally agree! :)

File: packager/media/formats/mp4/low_latency_segment_segmenter.cc (hunk 9 above)
Comment (CaitlinOCallaghan) [reply packager/media/formats/mp4/low_latency_segment_segmenter.cc]: This has been addressed. :)

File: packager/media/formats/mp4/low_latency_segment_segmenter.cc (hunk 7 above)
Comment (CaitlinOCallaghan) [reply packager/media/formats/mp4/low_latency_segment_segmenter.cc]: To "append" data to the the HTTP file, we aren't using the curl append mode. Rather, the file is being created and held open for the duration of 1 segment. After a chunk is created, it is immediately written to the already open file. The chunk data is stored in the  `IoCache upload_cache_` and on `IoCache::Write(...)`, a `base::WaitableEvent write_event_` is signaled, which triggers the transmission of the data to the remote file. I explored using [`CURLOPT_APPEND`](https://curl.se/libcurl/c/CURLOPT_APPEND.html), but it is only compatible with FTP. What are your thoughts? :) You have much more expertise in this area than I do.

File: packager/mpd/base/representation.cc (hunk 6 above)
Comment (CaitlinOCallaghan) [reply packager/mpd/base/representation.cc]: Your understanding is correct! However, the `frame_duration = sample_duration` as seen in [SetSampleDuration](https://github.com/google/shaka-packager/blob/master/packager/mpd/base/representation.cc#L196).  Each chunk contains a single media sample, so `sample_duration = chunk_duration`.  I added a clarifying comment to help out here, since the relationships are not clear.

File: packager/app/packager_main.cc (hunk 1 above)
Comment (CaitlinOCallaghan) [reply packager/app/packager_main.cc]: I added a clarifying comment above the error log. 

Also, those commas hurt my eyes, brain, and soul. I think I was going for a ", clarification/specification," kinda thing but forgot the first comma? Yikes.

File: packager/file/file.cc (hunk 3 above)
Comment (CaitlinOCallaghan) [reply packager/file/file.cc]: The switch to POST was an intermediate step in the solution development process. The current LL-DASH actually implementation works with both PUT and POST, since we are just opening a file, writing to it, and then closing it. I just switched this back to `kPut` :) .

File: packager/file/http_file.cc (hunk 4 above)
Comment (CaitlinOCallaghan) [reply packager/file/http_file.cc]: The LL-DASH is going to be using PUT now, but I still switched this line to your suggestion of `&download_cache_` if it's `kGet`. Might as well clean it up to account POST and PUT while we are here :).

File: packager/app/mpd_flags.cc (hunk 2 above)
Comment (CaitlinOCallaghan) [reply packager/app/mpd_flags.cc]: A description has been added :)

File: packager/app/mpd_flags.cc [hunk 13]
```diff
@@ -75,3 +75,10 @@ DEFINE_bool(dash_force_segment_list,
             "content is huge and the total number of (sub)segment references "
             "is greater than what the sidx atom allows (65535). Currently "
             "this flag is only supported in DASH ondemand profile.");
+DEFINE_bool(is_low_latency_dash,
```
Comment (CaitlinOCallaghan) [inline packager/app/mpd_flags.cc]: Proposed flag description:

"If enabled, low latency DASH (LL-DASH) will be used. This will reduce overall latency by ensuring that the media segments are chunk encoded and delivered via an aggregating response. The combination of these features will ensure that overall latency can be decoupled from the segment duration. For low latency to be achieved, the output of the packager must be combined with a delivery system which can chain together a set of aggregating responses (such as chunked transfer encoding under HTTP/1.1 or a HTTP/2 or HTTP/3 connection), as well as a DASH client which understands availabilityTimeOffset and the throughput estimation and ABR challenges which arise when operating in the low latency regime."

File: packager/app/packager_main.cc [hunk 14]
```diff
@@ -330,9 +330,13 @@ base::Optional<PackagingParams> GetPackagingParams() {
...
 
   if (chunking_params.is_low_latency_dash && chunking_params.subsegment_duration_in_seconds > 0) {
-    LOG(ERROR) << "Fragment duration --fragment_duration, "
-                  "cannot be specified if LL-DASH --is_low_latency_dash, "
-                  "is enabled.";
+    // Low latency streaming requires data to be shipped as chunks,
+    // the smallest unit of video. Therefore, in low latency mode,
+    // each fragment will only contain one chunk, defaulting the
+    // fragment duration to the shortest time possible
+    // and making a user specified --framgnet_dration irrelevant.
```
Comment (joeyparrish) [inline packager/app/packager_main.cc]: Typo: fragment_duration

File: packager/app/mpd_flags.cc (hunk 13 above)
Comment (CaitlinOCallaghan) [reply packager/app/mpd_flags.cc]: Moving the thorough description to docs/source/options/dash_options.rst (https://google.github.io/shaka-packager/html/tutorials/dash.html) and will include usage examples.

File: packager/media/formats/mp4/low_latency_segment_segmenter.cc [hunk 15]
```diff
@@ -0,0 +1,210 @@
...
+
+  UpdateProgress(GetSegmentDuration());
+
+  return Status::OK;
+}
+
+Status LowLatencySegmentSegmenter::FinalizeSegment() {
+  // Close the file now that the final chunk has been written
+  LOG(INFO) << "finalizing segment";
+  if (!segment_file_.release()->Close()) {
```
Comment (joeyparrish) [inline packager/media/formats/mp4/low_latency_segment_segmenter.cc:183]: After release(), the caller would have the job of destroying the object.  ...  And **OMG**, Close() calls `delete this`, which is a horrifying pattern that is not your fault.  😭

**Review by joeyparrish — CHANGES_REQUESTED** [review]
Looks great!

File: packager/app/packager_main.cc (hunk 14 above)
Comment (CaitlinOCallaghan) [reply packager/app/packager_main.cc]: Oh dear!

File: packager/media/formats/mp4/low_latency_segment_segmenter.cc (hunk 15 above)
Comment (CaitlinOCallaghan) [reply packager/media/formats/mp4/low_latency_segment_segmenter.cc:183]: I ran into some compiler errors of "calling a protected destructor of class 'shaka::File'" because of the File implementation. I learned lot about protected destructors and architecture on the side 😅.

File: packager/app/packager_main.cc [hunk 16]
```diff
@@ -334,7 +334,7 @@ base::Optional<PackagingParams> GetPackagingParams() {
     // the smallest unit of video. Therefore, in low latency mode,
     // each fragment will only contain one chunk, defaulting the
     // fragment duration to the shortest time possible
-    // and making a user specified --framgnet_dration irrelevant.
+    // and making a user specified --fragment_dration irrelevant.
```
Comment (joeyparrish) [inline packager/app/packager_main.cc]: Still missing a "u" in "duration"

**Review by joeyparrish — APPROVED** [review]

File: packager/app/packager_main.cc [hunk 17]
```diff
@@ -325,9 +325,21 @@ base::Optional<PackagingParams> GetPackagingParams() {
...
+  chunking_params.is_low_latency_dash = FLAGS_is_low_latency_dash;
   chunking_params.segment_sap_aligned = FLAGS_segment_sap_aligned;
   chunking_params.subsegment_sap_aligned = FLAGS_fragment_sap_aligned;
 
+  if (chunking_params.is_low_latency_dash && chunking_params.subsegment_duration_in_seconds > 0) {
+    // Low latency streaming requires data to be shipped as chunks,
+    // the smallest unit of video. Therefore, in low latency mode,
+    // each fragment will only contain one chunk, defaulting the
+    // fragment duration to the shortest time possible
+    // and making a user specified --fragment_duration irrelevant.
```
Comment (CaitlinOCallaghan) [inline packager/app/packager_main.cc:338]: Fixed the typo

File: packager/app/packager_main.cc (hunk 16 above)
Comment (CaitlinOCallaghan) [reply packager/app/packager_main.cc]: Fixed! :) https://github.com/CaitlinOCallaghan/shaka-packager/pull/3/files#r681359831

File: packager/media/formats/mp4/low_latency_segment_segmenter.cc (hunk 8 above)
Comment (CaitlinOCallaghan) [reply packager/media/formats/mp4/low_latency_segment_segmenter.cc]: No worries! I had to hunt down the buffer clearing. A lot of stuff is quite hidden - like the `delete` in `File::Close( )` 😅.

File: packager/app/packager_main.cc (hunk 17 above)
Comment (ngarg17) [reply packager/app/packager_main.cc:337]: Can a chunk be a frame?

Comment (ngarg17) [reply packager/app/packager_main.cc:338]: Can a chunk be a frame?

File: packager/file/http_file.cc [hunk 18]
```diff
@@ -297,7 +297,7 @@ void HttpFile::SetupRequest() {
   curl_easy_setopt(curl, CURLOPT_FOLLOWLOCATION, 1L);
   curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, &CurlWriteCallback);
   curl_easy_setopt(curl, CURLOPT_WRITEDATA,
-                   method_ == HttpMethod::kPut ? nullptr : &download_cache_);
+                   method_ == HttpMethod::kGet ? &download_cache_ : nullptr);
```
Comment (ngarg17) [inline packager/file/http_file.cc:300]: What's the reason for this change?

File: packager/media/formats/mp4/low_latency_segment_segmenter.cc [hunk 19]
```diff
@@ -0,0 +1,210 @@
...
+  RETURN_IF_ERROR(fragment_buffer()->WriteToFile(segment_file_.get()));
+
+  UpdateProgress(GetSegmentDuration());
+
+  return Status::OK;
+}
+
+Status LowLatencySegmentSegmenter::FinalizeSegment() {
+  // Close the file now that the final chunk has been written
+  LOG(INFO) << "finalizing segment";
```
Comment (ngarg17) [inline packager/media/formats/mp4/low_latency_segment_segmenter.cc]: Should this be a VLOG?

File: packager/media/formats/mp4/mp4_muxer.cc [hunk 20]
```diff
@@ -294,6 +295,9 @@ Status MP4Muxer::DelayInitializeMuxer() {
   if (options().segment_template.empty()) {
```
Comment (ngarg17) [inline packager/media/formats/mp4/mp4_muxer.cc:299]: Do you know what this option segment_template.empty() mean? I hope there is no case where options().segment_template is empty but mp4_params. is_low_latency_dash is enabled?

File: packager/mpd/base/representation.cc [hunk 21]
```diff
@@ -208,6 +208,13 @@ void Representation::SetSampleDuration(uint32_t frame_duration) {
   }
 }
 
+void Representation::SetSegmentDuration() {
+  int64_t sd = mpd_options_.mpd_params.target_segment_duration * media_info_.reference_time_scale();
+  if (sd <= 0)
+    return;
```
Comment (ngarg17) [inline packager/mpd/base/representation.cc:215]: Do you need to log an error here? Or this error case can be expected?

File: packager/mpd/base/xml/xml_node_unittest.cc [hunk 22]
```diff
@@ -368,13 +368,14 @@ TEST_F(LiveSegmentTimelineTest, OneSegmentInfo) {
   const uint64_t kStartTime = 0;
   const uint64_t kDuration = 100;
   const uint64_t kRepeat = 9;
+  const bool kIsLowLatency = false;
```
Comment (ngarg17) [inline packager/mpd/base/xml/xml_node_unittest.cc:371]: Looks like you have disabled low latency mode for UT. Is it not possible to add some test cases for low latency mode?

**Review by ngarg17 — APPROVED** [review]
Looks great from my side. Great work @CaitlinOCallaghan. Just added some minor comments and clarifying questions.

File: packager/app/packager_main.cc (hunk 17 above)
Comment (CaitlinOCallaghan) [reply packager/app/packager_main.cc:338]: Yes, a chunk can contain a single frame up to a handful of frames. In this case, each chunk contains one frame.

File: packager/mpd/base/xml/xml_node_unittest.cc (hunk 22 above)
Comment (CaitlinOCallaghan) [reply packager/mpd/base/xml/xml_node_unittest.cc:371]: The unit tests are coming up soon! :)

File: packager/file/http_file.cc (hunk 18 above)
Comment (CaitlinOCallaghan) [reply packager/file/http_file.cc:300]: This change was a part of a [discussion I had with Joey](https://github.com/CaitlinOCallaghan/shaka-packager/pull/3#discussion_r679521915). For background, POST was considered as an upload option when I was fiddling with the LL-DASH implementation; however, the final implementation uses PUT. 

The [POST method is supported by Packager](https://github.com/google/shaka-packager/blob/master/packager/file/http_file.h#L25), but is not currently used by any feature. While looking at the `http_file.cc`, we noticed that this line's original logic did not account for POST. To support POST, which would need the `nullptr` value,  the ternary operator logic was changed.

File: packager/media/formats/mp4/low_latency_segment_segmenter.cc (hunk 19 above)
Comment (CaitlinOCallaghan) [reply packager/media/formats/mp4/low_latency_segment_segmenter.cc]: Nice catch! Both LOG(INFO) and VLOG are used in Packager, but I think VLOG is preferred. However, this comment was actually more for myself while I was writing the feature, so I am going to remove it.

File: packager/mpd/base/representation.cc (hunk 21 above)
Comment (CaitlinOCallaghan) [reply packager/mpd/base/representation.cc:215]: Good call. This case would only be hit if the segment duration is negative or 0. It seems that Packager does not check for negative segment durations anywhere else, which is an issue. I'll add an error to address this.

File: packager/media/formats/mp4/mp4_muxer.cc (hunk 20 above)
Comment (CaitlinOCallaghan) [reply packager/media/formats/mp4/mp4_muxer.cc:299]: This case will hit if the user does not specify a segment template for the DASH manifest. I don't see a case where you'd want ULL without a segment template, but I am going to verify this with someone who is well versed in DASH protocol, and then I will adjust the logic if needed.

**Review by stephenlindauer — APPROVED** [review]
